src/sbuf.c

```diff
@@ -1235,6 +1236,44 @@ static bool setup_tls(struct tls_config *conf, const char *pfx, int sslmode,
 	return true;
 }
 
+static bool tls_config_changed(struct tls_config *new_server_connect_conf)
+{
+	return (strings_equal(new_server_connect_conf->ca_file, server_connect_conf->ca_file) &&
+		strings_equal(new_server_connect_conf->ca_path, server_connect_conf->ca_path) &&
+		strings_equal(new_server_connect_conf->ca_mem, server_connect_conf->ca_mem) &&
+		(new_server_connect_conf->ca_len == server_connect_conf->ca_len) &&
+		strings_equal(new_server_connect_conf->ciphers, server_connect_conf->ciphers) &&
+		(new_server_connect_conf->ciphers_server == server_connect_conf->ciphers_server) &&
+		(new_server_connect_conf->dheparams == server_connect_conf->dheparams) &&
+		(new_server_connect_conf->ecdhecurve == server_connect_conf->ecdhecurve) &&
```

Comment:
Ah yes, it was on purpose but not necessarily for "good" reasons. See https://github.com/pgbouncer/pgbouncer/pull/1157#discussion_r1741782953

It can be worked through perhaps in the libusual PR.

Reply:
@JelteF I added some code to do this and a test that fails on [another branch here](https://github.com/pgbouncer/pgbouncer/commit/3e3fca58df9b50281ab2307aa3016911dbbccdd2).

Breakpointing in gdb it would appear that these sub-structs are NULLed out/not hydrated by libusual on config load... on the branch linked above:

```
(gdb) break sbuf.c:1274
Breakpoint 3 at 0x55555557dda3: file src/sbuf.c, line 1274.
(gdb) c
Continuing.
2024-09-04 14:16:43.208 UTC [1003231] LOG S-0x55555566ab80: pTxnPool/bouncer@127.0.0.1:10201 new connection to server (from 127.0.0.1:37876)
2024-09-04 14:16:48.748 UTC [1003231] LOG RELOAD command issued

Breakpoint 3, tls_config_unchanged (new_server_connect_conf=0x55555567e930)
    at src/sbuf.c:1274
1274            return (strings_equal(new_server_connect_conf->ca_file, server_connect_conf->ca_file) &&
(gdb) print new_server_connect_conf[0].keypair[0]
$22 = {next = 0x0, cert_file = 0x0, cert_mem = 0x0, cert_len = 0, key_file = 0x0, key_mem = 0x0, key_len = 0}
(gdb) print server_connect_conf[0].keypair[0]
$23 = {next = 0x0, cert_file = 0x0, cert_mem = 0x0, cert_len = 0, key_file = 0x0, key_mem = 0x0, key_len = 0}
(gdb)```

It's possible that these only get values when libusual needs to read in those files and create these objects.

I'll add a code comment and at this point it's time to move the work over to libusual.

Reply:
[1c373c5](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/1c373c58937bb5a376fba736cb093ceacaea3e81)